### ISSUE TITLE
Payment proposals should lock the smallest amount possible.

### ISSUE THREAD

**Issue by marvin8** [bitpay/copay#3473]
I have a 2-of 2 miltu sig wallet with a mate that we use as cheap escrow.
One issue we've run into a couple of times now is that creating a payment proposal often ends up locking up way more funds than we would expect

See screenshot where a payment proposal of 1.2 BTC ends up locking over 2 BTC:
![screenshot from 2015-11-18 13-13-27](https://cloud.githubusercontent.com/assets/10649901/11232143/6846db1e-8dfb-11e5-8987-3526e0991266.png)

Looking at balances for each address currently in the wallet (see below) copay would have been able to only lock 1.35 BTC for this payment proposal instead.
![screenshot from 2015-11-18 13-13-08](https://cloud.githubusercontent.com/assets/10649901/11232171/9f1690f8-8dfb-11e5-83fd-3efbcf1a5ade.png)

This is a real issue for us and is becoming quite a nuisance. Most of our payment proposals stay in the wallet for a few days until the counterparty is happy to release funds, so in our case this is not just a short term issue but an issue that is seriously affecting how we can use copay.

Right now, this wallet has a balance of nearly 3 BTC and we need to create another payment proposal of ~ 1.2 BTC which on the surface should be possible (1.2 + 1.2 < 3) but the way copay has locked funds, we cannot create this second payment proposal as copay as locked up over 2 BTC already.

(might be a duplicate of issue #2545 and implementing unfunded proposals as per #169 might help in our situation)

**Comment by matiu** [comment]
Hi,

There are different criteria for the *best* input selection. Privacy, security and convenience need to be considered. Currently Copay prioritize confirmed inputs with 6+ confirmations, then inputs with 1+ confirmation and the unconfirmed inputs (you can select not to use unconfirmed inputs).

What would you think if input selection is selectable, as mentioned here: https://github.com/bitpay/copay/issues/2545#issuecomment-157600139

**Comment by dabura667** [comment]
tbh, I think it would be MORE work to have 3 selections and make some logic for 3 different coin selection algorithms than it would be just to have an advanced menu where all utxos are listed (with a "locked" tag on one's currently in proposal) and the user can tap them to use them.

Anyone who cares about utxos can then select them at their own leisure. (plus it is probably easier to code)

However, I do like the proposal to have a preliminary proposal with no specified inputs, and:

1. If the balance of the wallet falls below the amount proposed to be sent, the proposal is immediately disposed of.
2. After the proposal gets m signatures, the transaction moves into "Pending signatures" state. Next time any wallet that approved the proposal connects to BWS, the Copay app automatically signs the transaction. (This might be tricky, as it seems that if done incorrectly could be an attack vector... or you can re-confirm with them like "We will now sign proposal # 4 with your key (which you approved on Wednesday @ 4pm) Is this ok?" etc.)

2FA people will use the old proposal style, and multi-sig using businesses don't need instant broadcasting anyways (since they are likely waiting for approvals anyways).

**Comment by marvin8** [comment]
Just as further input, when the first payment proposal was entered all inputs, including change inputs, would have had 6+ confirmations. Maybe its just a case that inputs with 6+ confirmations could be used in such a way to minimise the locked amount.

I'd be happy if there was an advanced setting allowing us to pick what algorithm / policy is used to select inputs to be locked for a payment proposal.
At this stage, anything that makes payment proposals act a little more as expected would be very much appreciated.

By expected behavior I mean, that a layman's expectation would surely be that a payment proposal of e.g. 1.2 BTC locks 1.2 BTC in the wallet, regardless of what the inputs look like. For users with a non-tech background the locking based on inputs would be confusing.

**Comment by dan-da** [comment]
Marvin makes a good point.  Copay (and all software) should aim for least user surprise.
See: https://en.wikipedia.org/wiki/Principle_of_least_astonishment

Regardless of whatever the present selection algorithm is, and even if the user chose the algorithm, the user could still be surprised if it acts differently from his or her (possibly naive) expectation.

In my mind, the least surprise would occur if the user is provided an option to preview which utxo's will be used and then confirm/abort.  Not every user will care or wish to do this, so it should be optional.

**Comment by marvin8** [comment]
So I have just run into this again... This time I kind of expected it, but it's still kind of confusing at first even for tech heads... :-)

This time, my mate made a deposit of around 6 BTC in one payment.
No I'd like to create 4 payment requests of between 1.2 and 1.5 BTC each. Again, on the surface that looks like it should work :smile: 

However the first payment request for ~ 1.5 BTC has locked the full 6 BTC which means I can't create the other payment requests... Rather annoying!

Is there a way to lock part of the balance without actually locking the BTC in an individual address completely?

**Comment by dabura667** [comment]
I think this issue should be renamed to "proposals shouldn't lock funds"

If you send 6 BTC in, and make ANY proposal the current way proposals work will ALWAYS lock the entire 6 BTC if that is your only utxo available.

It is very non-intuitive, and the way proposals work should be revamped for multisig imo.

ie. Separate current proposal into two separate steps:
1. "Proposals" don't lock funds, and merely suggest amounts and outpoints. Once this receives m signatures, it is approved.
2. "Spend requests" lock funds, but can contain multiple approved outpoint/amount pairs.

So in OPs case, he would be able to generate 3 separate proposals and get approval before making the transaction. Then generate the transaction for signing with all approved outpoints and get a second round of approvals (which actually signs the transaction)

tbh this is better usability wise, as most environments that require waiting on approvals won't mind on waiting for two rounds of approvals rather than waiting for 6 rounds of approvals (propose, wait for approval, propose, wait for approval x 6 as compared to propose 6 proposals, wait for all 6 approvals, then generate 1 transaction with all 6 outpoints and get that signed.)

**Comment by dabura667** [comment]
Previously discussed "competing proposals" proposal is also inadequate, as if you make 6 proposals on a 6 BTC utxo and want to send 0.99 BTC to 6 different addresses, and all of those are approved by the quorum, it should send the utxo to 6 outpoints. Competing proposals would require the long waiting times between proposals just as I explained before.

**Comment by marvin8** [comment]
I'd be happy if there was a setting to disable looking of funds for proposals :smile:

**Comment by dabura667** [comment]
> I'd be happy if there was a setting to disable looking of funds for proposals

This would not fix your situation described below:

> However the first payment request for ~ 1.5 BTC has locked the full 6 BTC which means I can't create the other payment requests... Rather annoying!

Here's why:

1. Imagine when someone sent 6 BTC into your Copay wallet, they sent you a 6 dollar bill (imagine they exist) with a unique serial number on it.
2. A "proposal" in Copay is saying "We should spend the 6 dollar bill with serial no. 123 by giving 1.5 dollars to xyz and returning the remainder to ourselves." and the only thing the user can decide here is "giving 1.5 dollars to xyz"
3. The "locking" function is to say: "ok, while the group is arguing about how to spend the 6 dollar bill with serial no. 123 let's make sure no one else can sign  it to somewhere else.

So if we disabled locking, this is what you would see:

1. You propose 1.5 to xyz.
2. You propose 1.6 to abc.
3. You propose 1.9 to def.
4. Your mate approves 1.6 to abc.
5. Because "xyz" and "def" sends both refer to the same 6 BTC with the same unique ID. You must re-make the proposals by referencing the 4.4 BTC change you sent back to yourself when sending to abc and sending to xyz and def, otherwise it is seen as a "double spend" by the network.
6. However, if we don't reference the 6 BTC with a unique ID, then it will require two rounds of approvals. One for spend approval, and one for actually generating the bitcoin transaction. The benefits to this is that we can now make a transaction that says "We should spend the 6 dollar bill with serial no. 123 by giving 1.5 dollars to xyz, 1.6 dollars to abc, 1.9 dollars to def, and returning the remainder to ourselves." which would mean less time with funds "locked"... however, more time approving things, and must open the Copay app twice instead of once to approve.

**Comment by marvin8** [comment]
Thanks for explaining that... Now I am thinking there's not really much that can be done, without reverting back to a centralised type of wallet... and I don't really want to go there :-)

**Comment by n13** [comment]
@dabura667 Sort of makes sense given the limitations but there is no reason other accounts could not be added to the request and it would work the same way. 

So ... add 27 recipients, send it off to be signed by the other signers.. then it's saying "amount x goes to A, amount y goes to B, amount z goes to C, .... ... and the whole rest goes back to us" << valid BTC transaction
